src/cudadev/CUDACore/CachingDeviceAllocator.h

```diff
@@ -445,17 +455,17 @@ namespace notcub {
           BlockDescriptor free_key(device);
           CachedBlocks::iterator block_itr = cached_blocks.lower_bound(free_key);
 
+          // Store the list of memory blocks and events to release (this will be done
+          // outside of the critical section
+          std::vector <BlockAndEvent> toFree;
+
           while ((block_itr != cached_blocks.end()) && (block_itr->device == device)) {
             // No need to worry about synchronization with the device: cudaFree is
             // blocking and will synchronize across all kernels executing
             // on the current device
 
-            // Free device memory and destroy stream event.
-            // CMS: silently ignore errors and pass them to the caller
-            if ((error = cudaFree(block_itr->d_ptr)))
-              break;
-            if ((error = cudaEventDestroy(block_itr->ready_event)))
-              break;
+            // CMS: Free device memory and destroy stream event later.
+            toFree.push_back(BlockAndEvent{block_itr->d_ptr, block_itr->ready_event});
```

Comment:
Note that this happens only if `cudaMalloc()` fails because of no free memory, i.e. it should not happen normally.